Restrict `setuptools` to 75.8.0 as maximum when building a package.

mauritsvanrees opened the pull request:
Require `wheel` as well in the `build-system`.
Explicitly set the build backend to use `setuptools`.

See also [pep 517](https://peps.python.org/pep-0517/) and [pep 518](https://peps.python.org/pep-0518/).

Fixes https://github.com/plone/meta/issues/267

Used in https://github.com/plone/plone.app.caching/pull/148

mauritsvanrees committed:
Restrict `setuptools` to 75.8.0 as maximum when building a package.

Require `wheel` as well in the `build-system`.
Explicitly set the build backend to use `setuptools`.

See also [pep 517](https://peps.python.org/pep-0517/) and [pep 518](https://peps.python.org/pep-0518/).

Fixes https://github.com/plone/meta/issues/267

mauritsvanrees commented:
I think this is fine and could help.  Tried in `plone.app.caching` in coredev 6.2.  Then I called `../../bin/zopepy -m build .`  Result:

```
$ ll dist/
-rw-r--r--  1 maurits  staff  104030  6 Mar 11:02 plone.app.caching-4.1.0-py3-none-any.whl
-rw-r--r--  1 maurits  staff   95810  6 Mar 11:02 plone_app_caching-4.1.0.tar.gz
```

That is what I expect, and this is good.  BTW, you will get a warning by email from PYPI when you upload such a wheel.  But for now it is good.

For some reason though, zest.releaser gives problems.  In coredev locally I am using a checkout of zest.releaser, because the latest wheel has the naming problem that we are trying to avoid here.  But running it gives yet again a problem:

```
$ ../../bin/fullrelease 
INFO: Starting prerelease.
Cannot cleanup history, will try again later.
Run pyroma on the package before tagging? (Y/n)? 
...
Do you want to run check-manifest? (Y/n)? 
['/Users/maurits/community/plone-coredev/6.2/bin/python3.13', '-m', 'build', '--sdist', '.', '--outdir', '/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/check-manifest-qhyqv6_e-sdist'] failed (status 1):
/Users/maurits/community/plone-coredev/6.2/bin/python3.13: No module named build

Something bad happened. Do you want to continue despite that? (y/N)? 
```

Okay, this is actually a problem in check-manifest, which expects that the `build` tool is available in `bin/python`, which it is not.  Let's park that problem for now, and skip the check-manifest check:

```
$ ../../bin/fullrelease 
...
INFO: Tag checkout placed in /private/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/plone.app.caching-4.1.0-whele5pw/gitclone
INFO: Changing to sub directory in tag checkout: /private/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/plone.app.caching-4.1.0-whele5pw/gitclone
INFO: Finding and compiling po files in /private/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/plone.app.caching-4.1.0-whele5pw/gitclone
INFO: Making a source distribution of a fresh tag checkout (in /private/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/plone.app.caching-4.1.0-whele5pw/gitclone).
INFO: Creating isolated environment: venv+pip...
INFO: Installing packages in isolated environment:
- setuptools>=68.2,<=75.8.0
- wheel
INFO: Building sdist...
...
INFO: Making a wheel of a fresh tag checkout (in /private/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/plone.app.caching-4.1.0-whele5pw/gitclone).
INFO: Getting build dependencies for wheel...
...
INFO: Building wheel...
running bdist_wheel
running build
...
creating build/bdist.macosx-14.7-x86_64/wheel/plone_app_caching-4.1.0.dist-info/WHEEL
creating '/private/var/folders/26/1plvhxbs6yx7g_82v2xdxc500000gn/T/plone.app.caching-4.1.0-whele5pw/gitclone/dist/.tmp-hltggrw6/plone_app_caching-4.1.0-py3-none-any.whl' and adding 'build/bdist.macosx-14.7-x86_64/wheel' to it
...
Upload to pypi (Y/n)? n
```

And at that point I answer no.  Because when you look in the `dist` folder of the tag checkout, you see this:

```
-rw-r--r--  1 maurits  staff  104034  6 Mar 10:59 plone_app_caching-4.1.0-py3-none-any.whl
-rw-r--r--  1 maurits  staff   95401  6 Mar 10:59 plone_app_caching-4.1.0.tar.gz
```

So maybe [this part of `zest.releaser`](https://github.com/zestsoftware/zest.releaser/blob/9.4.0/zest/releaser/release.py#L160-L168) does not call the `build` tool correctly.

jensens commented:
For now fine.
In  the long run we should move to hatchling or to the upcoming feature in uv (currently experimental) as build backend.

mauritsvanrees committed:
pyproject.toml: do not set the build-backend.

This gives a problem with check-manifest, see https://github.com/mgedmin/check-manifest/issues/172

mauritsvanrees commented:
I actually saw the problem with check-manifest already last year.  See https://github.com/mgedmin/check-manifest/issues/172

Workaround: do not set a `build-backend` after all.

This does *not* fix the problem I have with zest.releaser though.

mauritsvanrees commented:
For good measure I tried with the same build-system as Zope uses:

```
[build-system]
requires = ["setuptools <= 75.6.0"]
build-backend = "setuptools.build_meta"
```

But this does not have the effect that I want.  Still a wheel with underscores is generated.

gforcada commented on pyproject.toml:
```diff
@@ -1,6 +1,5 @@
 [build-system]
-requires = ["setuptools < 74"]
-build-backend = "setuptools.build_meta"
+requires = ["setuptools>=68.2,<=75.8.0", "wheel"]
```
somehow this breaks `check-manifest`...

As this is used to build `plone.meta`, we might not need to get setuptools fixed here? 🤔

mauritsvanrees replied:
Ah, without build-backend it fails because then we get the default build-backend, which always calls `setup.py`, which we don't have.  Okay, I have restored that line.

I think the upper bound is still useful.  Otherwise when you install plone.meta using latest buildout, the result may have the same problems.

mauritsvanrees committed:
In our own pyproject.toml we must set a build-backend

mauritsvanrees commented:
Still no clue why `../../bin/fullrelease` would result in building a wheel with underscores (`plone_app.caching...whl`), where `../../bin/zopepy -m build .` results in a wheel with dots.  So the change in this PR has no effect when you use zest.releaser...

And I have a [fix in buildout](https://github.com/buildout/buildout/pull/687) so that it generates a properly named egg which actually works.  So this PR may be less needed.

mauritsvanrees committed:
Remove the upper bound of setuptools in the build-system again.

zc.buildout 4.1.4 can handle it.

mauritsvanrees commented:
With `zc.buildout` 4.1.4 the upper bound on `setuptools` should no longer be needed.  Also, if we use this older setuptools, you will get a warning email from PyPI that you are uploading a wheel with a legacy name, which at some point in the future will become an error.

So I suppose it is fine to leave the upper bound away.  I will push a commit for that.  I have made a few package releases now, will push those shortly to coredev.

BTW, I have made the releases with the `zest.releaser` branch from [this PR](https://github.com/zestsoftware/zest.releaser/pull/463).  This now tells you which files are going to be uploaded, before it asks you if you want to upload.  For example when releasing `plone.autoinclude` just now:

```
Current working directory: /private/var/folders/.../plone.autoinclude-2.0.1-sptkwq_g/gitclone
These files are ready for upload:
- dist/plone_autoinclude-2.0.1-py3-none-any.whl
- dist/plone_autoinclude-2.0.1.tar.gz
```

While releasing `plone.app.caching`, which has the upper bound of setuptools from the current PR, that zest.releaser version now correctly uses the upper bound and generated "old-style" wheels with a dot:

```
These files are ready for upload:
- dist/plone.app.caching-4.1.0-py3-none-any.whl
- dist/plone_app_caching-4.1.0.tar.gz
```

So that problem in zest.releaser will be solved after that is merged.

mauritsvanrees commented:
> So I suppose it is fine to leave the upper bound away. I will push a commit for that. I have made a few package releases now, will push those shortly to coredev.

Apparently, I only updated the `pyproject.toml` of `plone.meta` itself, _not_ the `pyproject.toml.j2` of generated projects.  I have done so now in https://github.com/plone/meta/pull/275.  But there I set the build-system to stay below setuptools 77.
